# Bank Customer Segmentation 

In [1]:
#importing libraries
import numpy as np
import pandas as pd 
import warnings

In [2]:
#reading file
df=pd.read_csv("Desktop/data/bank_transactions.csv")
df.head()

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053     10/1/94          F   JAMSHEDPUR   
1            T2   C2142763      4/4/57          M      JHAJJAR   
2            T3   C4417068    26/11/96          F       MUMBAI   
3            T4   C5342380     14/9/73          F       MUMBAI   
4            T5   C9031234     24/3/88          F  NAVI MUMBAI   

   CustAccountBalance TransactionDate  TransactionTime  \
0            17819.05          2/8/16           143207   
1             2270.69          2/8/16           141858   
2            17874.44          2/8/16           142712   
3           866503.21          2/8/16           142714   
4             6714.43          2/8/16           181156   

   TransactionAmount (INR)  
0                     25.0  
1                  27999.0  
2                    459.0  
3                   2060.0  
4                   1762.5

# Data Cleaning

In [15]:
df.columns

Index(['TransactionID', 'CustomerID', 'CustomerDOB', 'CustGender',
       'CustLocation', 'CustAccountBalance', 'TransactionDate',
       'TransactionTime', 'TransactionAmount', 'CoustomerAge'],
      dtype='object')

In [4]:
df.describe()

CustAccountBalance  TransactionTime  TransactionAmount (INR)
count        1.046198e+06     1.048567e+06             1.048567e+06
mean         1.154035e+05     1.570875e+05             1.574335e+03
std          8.464854e+05     5.126185e+04             6.574743e+03
min          0.000000e+00     0.000000e+00             0.000000e+00
25%          4.721760e+03     1.240300e+05             1.610000e+02
50%          1.679218e+04     1.642260e+05             4.590300e+02
75%          5.765736e+04     2.000100e+05             1.200000e+03
max          1.150355e+08     2.359590e+05             1.560035e+06

In [5]:
#checking for N/A values
df.isna().sum()

TransactionID                 0
CustomerID                    0
CustomerDOB                3397
CustGender                 1100
CustLocation                151
CustAccountBalance         2369
TransactionDate               0
TransactionTime               0
TransactionAmount (INR)       0
dtype: int64

In [6]:
#droping N/A values
df = df.dropna()
df.isna().sum()

TransactionID              0
CustomerID                 0
CustomerDOB                0
CustGender                 0
CustLocation               0
CustAccountBalance         0
TransactionDate            0
TransactionTime            0
TransactionAmount (INR)    0
dtype: int64

In [7]:
#to find null values
df.isnull().sum()

TransactionID              0
CustomerID                 0
CustomerDOB                0
CustGender                 0
CustLocation               0
CustAccountBalance         0
TransactionDate            0
TransactionTime            0
TransactionAmount (INR)    0
dtype: int64

In [8]:
#to find abnormal pattern in dates
df['CustomerDOB'].value_counts()

1/1/1800    56292
1/1/89        809
1/1/90        784
6/8/91        698
1/1/91        665
            ...  
13/11/39        1
14/10/36        1
20/12/58        1
9/3/59          1
25/10/02        1
Name: CustomerDOB, Length: 17233, dtype: int64

Since there 56292 whose DOB are registered as 1800-01-01, we have a total data of 1Million plus bank transactions so we will discard them

In [9]:
#Removal of dates which are absurd.
df = df.loc[~(df['CustomerDOB'] == '1/1/1800')]
df['CustomerDOB'].value_counts()

1/1/89     809
1/1/90     784
6/8/91     698
1/1/91     665
1/1/92     631
          ... 
9/4/63       1
19/6/01      1
3/2/75       1
8/11/65      1
1/12/55      1
Name: CustomerDOB, Length: 17232, dtype: int64

# updating date formats

In [10]:
#assigning proper format
df['TransactionDate'] = pd.to_datetime(df['TransactionDate'], format = '%d/%m/%y')
df['CustomerDOB'] = pd.to_datetime(df['CustomerDOB'])

In [11]:
df.head()

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053  1994-10-01          F   JAMSHEDPUR   
1            T2   C2142763  2057-04-04          M      JHAJJAR   
2            T3   C4417068  1996-11-26          F       MUMBAI   
3            T4   C5342380  1973-09-14          F       MUMBAI   
4            T5   C9031234  1988-03-24          F  NAVI MUMBAI   

   CustAccountBalance TransactionDate  TransactionTime  \
0            17819.05      2016-08-02           143207   
1             2270.69      2016-08-02           141858   
2            17874.44      2016-08-02           142712   
3           866503.21      2016-08-02           142714   
4             6714.43      2016-08-02           181156   

   TransactionAmount (INR)  
0                     25.0  
1                  27999.0  
2                    459.0  
3                   2060.0  
4                   1762.5

# Customer age during the transaction

In [12]:
# conversion of years greater than 2021 
df.loc[df['CustomerDOB'].dt.year >= 2021, ['CustomerDOB']] -= pd.DateOffset(years = 100)
#calculation
df['CoustomerAge'] = (pd.to_datetime('today') - df['CustomerDOB'])/np.timedelta64(1, 'Y')

df.head()

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053  1994-10-01          F   JAMSHEDPUR   
1            T2   C2142763  1957-04-04          M      JHAJJAR   
2            T3   C4417068  1996-11-26          F       MUMBAI   
3            T4   C5342380  1973-09-14          F       MUMBAI   
4            T5   C9031234  1988-03-24          F  NAVI MUMBAI   

   CustAccountBalance TransactionDate  TransactionTime  \
0            17819.05      2016-08-02           143207   
1             2270.69      2016-08-02           141858   
2            17874.44      2016-08-02           142712   
3           866503.21      2016-08-02           142714   
4             6714.43      2016-08-02           181156   

   TransactionAmount (INR)  CoustomerAge  
0                     25.0     27.241874  
1                  27999.0     64.734772  
2                    459.0     25.087141  
3                   2060.0     48.288165  
4                   1762.5     33.763568

In [13]:
#renaming the column
df = df.rename(columns = {'TransactionAmount (INR)' : 'TransactionAmount'})

In [14]:
#convertig age to int type from float.
df=df.astype({'CoustomerAge':int})
df.head()

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053  1994-10-01          F   JAMSHEDPUR   
1            T2   C2142763  1957-04-04          M      JHAJJAR   
2            T3   C4417068  1996-11-26          F       MUMBAI   
3            T4   C5342380  1973-09-14          F       MUMBAI   
4            T5   C9031234  1988-03-24          F  NAVI MUMBAI   

   CustAccountBalance TransactionDate  TransactionTime  TransactionAmount  \
0            17819.05      2016-08-02           143207               25.0   
1             2270.69      2016-08-02           141858            27999.0   
2            17874.44      2016-08-02           142712              459.0   
3           866503.21      2016-08-02           142714             2060.0   
4             6714.43      2016-08-02           181156             1762.5   

   CoustomerAge  
0            27  
1            64  
2            25  
3            48  
4            33